ddtrace/utils/hook.py

```diff
@@ -0,0 +1,196 @@
+    """
+    def load_module(self, fullname):
+        module = sys.modules[fullname]
+        notify_module_loaded(module)
+        return module
+
+
+class _ImportHookChainedLoader(object):
+    def __init__(self, loader):
+        self.loader = loader
+
+    def load_module(self, fullname):
+        module = self.loader.load_module(fullname)
+        notify_module_loaded(module)
+        return module
+
+
+class ImportHookFinder:
+    def __init__(self):
+        self.in_progress = {}
+
+    @synchronized(_post_import_hooks_lock)
+    def find_module(self, fullname, path=None):
+        # If the module being imported is not one we have registered
+        # post import hooks for, we can return immediately. We will
+        # take no further part in the importing of this module.
+
+        if fullname not in _post_import_hooks:
+            return None
+
+        # When we are interested in a specific module, we will call back
+        # into the import system a second time to defer to the import
+        # finder that is supposed to handle the importing of the module.
+        # We set an in progress flag for the target module so that on
+        # the second time through we don't trigger another call back
+        # into the import system and cause a infinite loop.
+
+        if fullname in self.in_progress:
+            return None
+
+        self.in_progress[fullname] = True
+
+        # Now call back into the import system again.
+
+        try:
+            if PY3:
+                # For Python 3 we need to use find_spec().loader
+                # from the importlib.util module. It doesn't actually
+                # import the target module and only finds the
+                # loader. If a loader is found, we need to return
+                # our own loader which will then in turn call the
+                # real loader to import the module and invoke the
+                # post import hooks.
+                try:
+                    import importlib.util
+                    loader = importlib.util.find_spec(fullname).loader
+                except (ImportError, AttributeError):
+                    loader = importlib.find_loader(fullname, path)
+                if loader:
+                    return _ImportHookChainedLoader(loader)
+            else:
+                # For Python 2 we don't have much choice but to
+                # call back in to __import__(). This will
+                # actually cause the module to be imported. If no
+                # module could be found then ImportError will be
+                # raised. Otherwise we return a loader which
+                # returns the already loaded module and invokes
+                # the post import hooks.
+                __import__(fullname)
+                return _ImportHookLoader()
+
+        finally:
+            del self.in_progress[fullname]
+
+
+@synchronized(_post_import_hooks_lock)
+def deregister_post_import_hook(modulename, matcher):
```

Comment:
Remind me, why this approach vs:

```python
def hook():
    pass

register_post_import_hook('module_name', hook)
deregister_post_import_hook('module_name', hook)
```

Reply:
I think in my initial implementation I had the idea of attaching an 'enabled' property on the hooks that would be checked. The `matcher` would allow us to easily filter out the disabled hooks.

Reply:
ok, as long as it makes sense with the plan

Reply:
Yeah, I'm getting myself back up to speed to see if we still want to go in this direction.

Reply:
I opted to just accept the hook. If I recall correctly, our plan was to have a layer on top of this that would be meant for enabling and disabling hooks which can interface with this general api.